src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetGeneratorServiceImpl.java

```diff
@@ -49,25 +50,35 @@ private void validateDatasetMetadataAndRecords(DatasetMetadata datasetMetadata,
   private Set<Record> processRecordsAndRemoveDuplicates(DatasetMetadata datasetMetadata, List<ByteArrayInputStream> records) {
     return records.stream()
         .map(ByteArrayInputStream::readAllBytes)
```

Comment:
You can remove line 53 to 64 and simplify it as the following explanation.

```
.map(record -> {
          RecordEntity recordEntity = recordRepository.save(new RecordEntity(null, null, datasetMetadata.getDatasetId()));
          Long recordId = recordEntity.getId();
          return Record.builder()
              .recordId(recordId)
              .datasetId(datasetMetadata.getDatasetId())
              .datasetName(datasetMetadata.getDatasetName())
              .country(datasetMetadata.getCountry())
              .language(datasetMetadata.getLanguage())
              .content(record)
              .build();
        })
```
and move this part of code inside the function ```getOptionalRecordFromProcessorService``` 
```
final RecordEntity recordEntity = recordRepository.save(new RecordEntity(null, null, datasetMetadata.getDatasetId()));
final Long recordId = recordEntity.getId();
```
The intent of that method was to surround and catch the ```exception``` that was thrown by ```xmlRecordProcessorService``` reported on MET-4093
now you can you use the same method to protect the ```recordRepository.save``` when it throws an exception, this was causing bug in MET-4093. 

the final code would look like this (see below), and you will need to refactor unit tests as well. I have renamed the method ```getOptionalRecordFromProcessorService``` to ```getOptionalRecordRepositorySaveService```

```
private Set<Record> processRecordsAndRemoveDuplicates(DatasetMetadata datasetMetadata, List<ByteArrayInputStream> records) {
    return records.stream()
        .map(ByteArrayInputStream::readAllBytes)
        .map(recordItem -> getOptionalRecordRepositorySaveService(datasetMetadata, recordItem))
        .flatMap(Optional::stream)
        .collect(toSet());
  }

  private Optional<Record> getOptionalRecordRepositorySaveService(DatasetMetadata datasetMetadata, byte[] record) {
    try {
      final RecordEntity recordEntity = recordRepository.save(new RecordEntity(null, null, datasetMetadata.getDatasetId()));
      final Long recordId = recordEntity.getId();
      return Optional.of(Record.builder()
                               .recordId(recordId)
                               .datasetId(datasetMetadata.getDatasetId())
                               .datasetName(datasetMetadata.getDatasetName())
                               .country(datasetMetadata.getCountry())
                               .language(datasetMetadata.getLanguage())
                               .content(record)
                               .build());
    } catch (Exception repositoryException) {
      LOGGER.error("Failed to save record {} :: {} ", new String(record), repositoryException);
      return Optional.empty();
    }
  }
```

Reply:
I see that the `develop` branch has these changes (or most of them) that you mentioned. I'll merge the develop into this branch and apply these changes